Automattic/mShots
# Support output dimensions with local dev

This PR combines #18 & #22. Each of those PRs "works" as a standalone PR, and we can merge either of them, but in this case each PR makes the other easier to test, so in this case it's arguably easier to review the one big diff rather than the two little ones independantly.

~As a reviewer please use the PR(s) that you feel most comfortable with, including commenting/approving whatever you feel comfortable with. I'll keep the three branches in sync and the other two branches operating independantly until one of them gets merged and everything gets simple again, so you can approve just one of the bases if you like.~

There were a number of little fixes required after combining these branches together, and cleaning up the git history doesn't seem like a good use of time unless someone cares. If you'd like to review/approve one of the source branches separately let me know and happy to cherry-pick all the changes into the two source branches.

### Testing Instructions

- `npm install`
- `npm start`
- Navigate to something like http://127.0.0.1:8000/mshots/v1/https://public-api.wordpress.com/rest/v1/template/demo/rockfield/rockfield%3Ffont_headings%3DPlayfair%20Display%26font_base%3DFira%20Sans%26site_title%3DRockfield%26viewport_height%3D600%26language%3Dfr?vph=1000&screen_height=3000
- You'll be redirected to Wait a few seconds while the screenshot is generated (you can use `docker-compose logs` to watch for the output if you prefer)
- navigate to the url again

You can find examples of the URLs we care about on the `https://wordpress.com/new/style` page.
- go to `https://wordpress.com/new/style`
- open devtools network requests
- choose a design

This will get you a url like:
https://s0.wp.com/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Frest%2Fv1%2Ftemplate%2Fdemo%2Fmaywood%2Fmaywood%3Ffont_headings%3DPlayfair%2520Display%26font_base%3DFira%2520Sans%26site_title%3DMaywood%26viewport_height%3D700%26language%3Dde?vpw=1600&vph=1600&w=600&h=600&count=1

Replace `https://s0.wp.com/` with `127.0.0.1:8000/` to point it at your dev server and add the new `?screen_height=3600` param to mshots itself, giving us a url like:

127.0.0.1:8000/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Frest%2Fv1%2Ftemplate%2Fdemo%2Fmaywood%2Fmaywood%3Ffont_headings%3DPlayfair%2520Display%26font_base%3DFira%2520Sans%26site_title%3DMaywood%26viewport_height%3D700%26language%3Dde?vpw=1600&vph=1600&w=600&h=600&count=1&screen_height=3600

You can also point calypso at it by editing `wp-calypso/client/landing/gutenboarding/components/mshots-image/index.tsx`:
`const mshotsUrl = 'http://127.0.0.1:8000/mshots/v1/';
```diff
-    const mshotsUrl = 'https://s0.wp.com/mshots/v1/';
+    const mshotsUrl = 'http://127.0.0.1:8000/mshots/v1/';
```

The problem with this approach is that you'll most likely swamp your little dev server and get a bunch of timeouts resulting in 404s. Then mshots will cache those 404 images and return them until you tell it to try again by adding `&requeue=true` to the query, e.g.: http://127.0.0.1:8000/mshots/v1/https%3A%2F%2Fpublic-api.wordpress.com%2Frest%2Fv1%2Ftemplate%2Fdemo%2Fmaywood%2Fmaywood%3Ffont_headings%3DPlayfair%2520Display%26font_base%3DFira%2520Sans%26site_title%3DMaywood%26viewport_height%3D700%26language%3Dde?vpw=1600&vph=1600&w=600&h=600&count=1&screen_height=3600&requeue=true

Once you requeue the individual images in a separate tab you can refresh the design page to see them there.
